Rewrite execProofMethod

felixlinker opened the pull request:
This PR essentially rewrites `execProofMethod` and, following from that, applies some other changes. Overall, this PR should be "idempotent", i.e., Tamarin should function exactly as before. I personally find the code cleaner than before, but this is a matter of taste.

The intention behind this PR is to factor out changes to Tamarin that I did as part of a research project with @cascremers. They are necessary for that project, but we would like to check that they indeed do not change Tamarin's behavior. But still, I think that some benefit to the added clarity of the code (more deletions than insertions).

Some notes on the changes:
- I added a new proof method `Finished Result` that replaces `Solved`, `Contradiction` and `Unfinishable`. I think it makes sense to reflect in the type system which kind of methods mean that no further proving needs to be done.
- I do not check in `execProofMethod` whether the applied method is valid in the given constraint system. In most cases, these checks are redundant because they come from the system. Instead, I implemented `checkAndExecProofMethod` that is now used whenever one must check validity, e.g., upon loading a theory.
- I deleted the `annotateWith(Diff)Systems` functions. They performed tasks fully redundant with `execProofMethod`. First, you would apply a proof method and delete the resulting constraint system, then you would annotate it again.
- I now always simplify after applying a proof method. This makes the every now and then appearing "simplify" steps obsolete. To maintain backwards compatibility, I do *not* anymore return `Nothing` when simplify results in the same constraint system in `execProofMethod`. If I were to do that, applying a redundant simplify step (e.g., while loading older proofs) would invalidated the proof.
- `execProofMethod` now checks for all ways in which a proof could be finished. Therefore, `applyProverAtPath` need not "guess" anymore which other provers we maybe should try to apply.

Fast regression tests pass in about the same time as before on my machine. Only difference is that there are slightly fewer steps which results from the now dropped `simplify` steps.

I'd appreciate a thorough review and am curious about any feedback!

felixlinker commented:
PS: I strongly suggest to merge by squashing and rebasing all commits. The commits themselves don't compile individually. I keep commits separate to be able to merge things back together in the end.

cascremers commented:
We'll look asap, thanks!
Note the only options are "squash and merge" or "rebase and merge" -- and I presume you want the first of these.

niklasmedinger commented:
Nit: In the file `ProofMethods.hs` line ~195 the comments for `Unfinishable` and `Contradictory` are wrong. They should be swapped.

niklasmedinger commented:
(Maybe) nit: In `ProofMethods.hs` line 269 the function `equalReason` only handles the `Contradictory _` case explicitly and uses a variable for the remaining cases.

If we were to explicitly handle all cases of `Result`, future changes of `Result` (like adding or removing a case) will lead to a compiler warning/error. This allows future developers to make changes to `equalReason` should it be necessary.

E.g.: With the current version of the function, a future developer could add a variant to `Result` which should be handled like the `Contradictory _` case but `equalReason` will simply used the derived `Eq` to check for equality. If we handle all cases explicitly, the compiler would complain.

I know this comment is pretty nit-picky/pedantic given that this style of handling patterns is used throughout the project. However, I've recently started working on the code and found it pretty annoying that changes to existing data types don't always lead to compiler warnings/errors because of 'default' cases like this and I have to manually investigate if I have to change functions that use the changed data type.

niklasmedinger commented on lib/theory/src/Theory/Constraint/Solver/ProofMethod.hs:
```diff
@@ -485,31 +498,37 @@ rankGoals ctxt ranking tacticsList sys = case ranking of
-    contradiction c                    = (Contradiction (Just c), "")
+    execMethods = mapMaybe execMethod
+    execMethod (m, expl) = do
+      cases <- execProofMethod ctxt m sys
```
Shouldn't this be a call to `checkAndExecProofMethods` now that `execProofMethod` no longer checks whether the method is valid? Or do we rely on `execProofMethod` returning `Nothing` if the execution fails?

felixlinker replied:
`rankProofMethods` computes all applicable methods itself, i.e., they must be valid (or the computation is buggy, but this shouldn't be fixed by checking that they are valid imo). Thus, checking whether it's valid is not necessary. At this point in the code, it cannot come from an "external" source such as a loaded proof file.

niklasmedinger commented on lib/theory/src/Theory/Constraint/Solver/ProofMethod.hs:
```diff
@@ -183,18 +189,26 @@ unmarkPremiseG annGoal                        = annGoal
+    Solved
+  -- ^ A dependency graph was found.
+  | Contradictory (Maybe Contradiction)
+  -- ^ The proof cannot be finished (due to reducible operators in subterms or
```
Comments for `Contradictory` and `Unfinishable` should be swapped.

niklasmedinger reviewed (commented):
Hi Felix,

thanks a lot for your work! Much appreciated.

Since I only recently got into the code base, I can't give much feedback. I'm only wondering whether `execProofMethod` can now be infallible because `checkAndExecProofMethod` exists now.

There's a small copy-paste error regarding the comments of `Result`. Apart from that I think the PR looks good!

felixlinker commented on lib/theory/src/Theory/Constraint/Solver/ProofMethod.hs:
```diff
@@ -232,100 +246,94 @@ instance HasFrees DiffProofMethod where
 -- NOTE that the returned systems have their free substitution fully applied
 -- and all variable indices reset.
 execProofMethod :: ProofContext
                 -> ProofMethod -> System -> Maybe (M.Map CaseName System)
```
No, that doesn't work because getting the induction cases can still fail.

felixlinker commented:
> We'll look asap, thanks! Note the only options are "squash and merge" or "rebase and merge" -- and I presume you want the first of these.

I mean the first of these, yes, but technically, you could also squash commits without rebasing 😉

jdreier commented:
My quick merge seems not to have worked (although compilation fails on different file), I guess @felixlinker needs to help here.

felixlinker commented:
@jdreier Did a new merge!

felixlinker commented:
Not quite sure why the regression tests fail. Is it because of the reduced number of proof steps? This is to be expected. My PR drops intermediate simplify steps.

cascremers commented:
There seems to be an error occurring in the check log -- it doesn't look like it is just the changes in the number of steps.

felixlinker commented:
@cascremers Thanks for catching that! I went over all logs and found a difference in how the solver works on the Yubikey example (https://github.com/tamarin-prover/tamarin-prover/actions/runs/10262365847/job/28391889024?pr=650#step:9:2923). Bottom-line: The reason was that `simplifySystem` currently might be applicable again, although its documentation suggests this shouldn't be the case.

I rewrote `simplifySystem` in the latest commit such that it is always applied until there are no more changes. While I was at it, I de-duplicated the code w.r.t. diff systems.

Here's my reasoning of the changes. Previously, `simplifySystem` had the following structure:
```hs
simplifySystem = do
  go 
  if isdiff then steps else steps'
  where
    go = do
      changed <- simplify  -- apply changes until system does not change anymore
      if changed then go else return ()
```

My thinking was: If we can apply `simplifySystem` multiple times leading to new changes, these must come from `steps` and `steps'` as `go` applies simplification until the system does not change anymore. Previously, it was checked whether we can apply `simplifySystem` again by just just doing it and checking whether the resulting constraint system differs w.r.t. the current one (check for equality). I really disliked this implementation as the `go` function carefully already checks whether there are changes or not (and this is contrast to the documentation of `simplifySystem`).

My implementation of `simplifySystem` now has the following structure:
```hs
simplifySystem = go
  where
    go = do
      if isdiff then steps else steps'  -- I did some deduplication here
      changed <- simplify
      if changed then go else return ()
```

In other words, I moved the steps that were previously applied after the recursion into the recursion, applying them before the steps that can lead to changes. 

I cannot be sure whether my changes are correct. They are, if the order of `steps`/`steps'` w.r.t. to the simplification steps in `simplify` is irrelevant. I have evidence to support this hypothesis:
- Regression tests pass on my machine.
- Previously, we would also apply `steps`/`steps'` if `simplify` resulted in `Unchanged` (suggesting `steps`/`steps'` does not rely on anything done in `simplify`).
- The cases in the Yubikey theory where a contradiction would previously be found after applying `simplify` again now yield a contradiction immediately.

Would be good to have someone else think about whether my changes are correct. Hence the detailed explanation!

felixlinker commented:
Changes are not idempotent. Wait with merge. I'm on it.

felixlinker commented:
@cascremers @jdreier PR is ready! Now, changes in number of proof steps just comes down to `simplify` being applied less.

felixlinker commented:
Side note: I **strongly** suggest merging by rebasing and squashing. The commit history is a hot mess and will confuse more than it will help.

felixlinker commented:
Regression tests seem to take 12% longer. I tried to confirm this locally and couldn't. I ran `python3 regressionTests.py -r=5` both on `origin/develop` and on this branch.

Tests took 27 minutes on `origin/develop` and 26 minutes on this branch.

cascremers commented:
Ok, looking good then!

jdreier commented:
Can you update the regression targets before merging?

cascremers commented:
Fair point by @jdreier .

Note that I don't know how to re-open this PR at the moment.

felixlinker commented:
I think I have to open a new PR.
